### PR TITLE
Prefer builtin swap functions

### PR DESCRIPTION
**Pull request by thbeu**
Get rid of multiple definitions of SwapWord in favour of built-in swap functions.

### CODE REVIEW

**Comment by rouault** [comment]
GDAL has a whole series of byte swapping macros that can be inlined by the compiler, and could potentially be of interest here to put in the shapefil_private.h file, and not require an extra .c file: https://github.com/OSGeo/gdal/blob/master/port/cpl_port.h#L725

**Comment by thbeu** [comment]
OK, let me check them. Thanks.

**Comment by thbeu** [comment]
Some remarks:

- GCC 4.8 seems to be the first GCC version to work on unsigned types. GCC &GreaterEqual; 4.3 and < 4.8 work on signed types.
- The _SHP_SWAPXY macros could be nodiscard or should not be called by library functions.
- The SHP_SWAPXY macros do the actual in-place swap.

File: shpendian.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#define SHPLIB_NULLPTR NULL
+#endif
+
+/************************************************************************/
+/*                              SwapWord()                              */
+/*                                                                      */
+/*      Swap a 2, 4 or 8 byte word.                                     */
+/************************************************************************/
+
+void SwapWord(int length, void *wordP)
```
Comment (rouault) [inline shpendian.c:34]: do we still need that function?

Comment (thbeu) [reply shpendian.c]: It's gone.

**Comment by rouault** [comment]
I've tried to integrate the changes of this pull request to GDAL, and it causes test failures in tests involving .qix (shptree.c). I don't get those failures on latest master
```
 ogr/ogr_shape.py ✓✓✓✓✓                                                                                                                                                                       2% ▎         

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_ogr_shape_8 ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

shape_ds = <osgeo.ogr.DataSource; proxy of <Swig Object of type 'OGRDataSourceShadow *' at 0x7fe7ec3839c0> >

    def test_ogr_shape_8(shape_ds):
    
        shape_dir = shape_ds.GetDescription()
        shape_lyr = shape_ds.GetLayer(0)
    
        index_fname = shape_dir + "/tpoly.qix"
    
        shape_ds.ExecuteSQL("CREATE SPATIAL INDEX ON tpoly")
    
        assert os.access(index_fname, os.F_OK), "tpoly.qix not created"
    
        with ogrtest.spatial_filter(shape_lyr, "LINESTRING(479505 4763195,480526 4762819)"):
>           ogrtest.check_features_against_list(shape_lyr, "eas_id", [158])
E           AssertionError: Got only 0 features, not the expected 1 features.
E           assert None is not None

/home/even/gdal/gdal/build_cmake/autotest/ogr/ogr_shape.py:209: AssertionError

 ogr/ogr_shape.py ⨯✓                                                                                                                                                                          3% ▍         

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_ogr_shape_10 ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    def test_ogr_shape_10():
    
        shape_ds = ogr.Open("data/shp/testpoly.shp")
        shape_lyr = shape_ds.GetLayer(0)
    
        with ogrtest.spatial_filter(shape_lyr, -400, 22, -120, 400):
>           ogrtest.check_features_against_list(shape_lyr, "FID", [0, 4, 8])
E           AssertionError: Got only 0 features, not the expected 3 features.
E           assert None is not None

/home/even/gdal/gdal/build_cmake/autotest/ogr/ogr_shape.py:246: AssertionError

 ogr/ogr_shape.py ⨯                                                                                                                                                                           4% ▍         

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_ogr_shape_11 ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    def test_ogr_shape_11():
    
        shape_ds = ogr.Open("data/shp/testpoly.shp")
        shape_lyr = shape_ds.GetLayer(0)
    
        with ogrtest.attribute_filter(shape_lyr, "FID = 5"), ogrtest.spatial_filter(
            shape_lyr, -400, 22, -120, 400
        ):
            ogrtest.check_features_against_list(shape_lyr, "FID", [])
    
        with ogrtest.attribute_filter(shape_lyr, "FID = 4"), ogrtest.spatial_filter(
            shape_lyr, -400, 22, -120, 400
        ):
>           ogrtest.check_features_against_list(shape_lyr, "FID", [4])
E           AssertionError: Got only 0 features, not the expected 1 features.
E           assert None is not None

/home/even/gdal/gdal/build_cmake/autotest/ogr/ogr_shape.py:266: AssertionError

 ogr/ogr_shape.py ⨯✓              
```

**Comment by thbeu** [comment]
Silly me! Glad you tested. I can squash the last two commits once your tests succeed.

**Comment by rouault** [comment]
Tried latest commit and I get a hard crash in SHPWriteObject. This test writes polygons
```
python3: /home/even/gdal/gdal/ogr/ogrsf_frmts/shape/shpopen.c:1321: gdal_SHPWriteObject: Assertion `psObject->nSHPType == psSHP->nShapeType || psObject->nSHPType == SHPT_NULL' failed.
Fatal Python error: Aborted

Current thread 0x00007fd8ebecf740 (most recent call first):
  File "/home/even/gdal/gdal/build_cmake/swig/python/osgeo/ogr.py", line 1261 in Destroy
  File "/home/even/gdal/gdal/build_cmake/autotest/ogr/ogr_shape.py", line 343 in test_ogr_shape_13
```

**Comment by thbeu** [comment]
Hm, this is rather strange and I do not see the error right now. The write operation is probably right, but how did you create that SHPObject?

I noticed some other locations for double pointer dereferencing (but only for big endian).

**Comment by rouault** [comment]
> The write operation is probably right, but how did you create that SHPObject?

cf https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/shape/shape2ogr.cpp#L519

> I noticed some other locations for double pointer dereferencing (but only for big endian).

This is a little-endian build. I haven't tried it yet with big-endian CI (we do have one one GDAL)

**Comment by thbeu** [comment]
I am still rather clueless regarding your little-endian assertion. Do you think you could add some small unit test to shapelib, too, such that we detect early in CI and I can reproduce and debug?

i can easily "mock" big-endian compilation to get all these other compiler issues.

**Comment by rouault** [comment]
> Do you think you could add some small unit test to shapelib, too, such that we detect early in CI and I can reproduce and debug?

honestly, I don't have time to do that... I've enough on my plate with GDAL.

**Comment by rouault** [comment]
I've given a try at this PR with GDAL in https://github.com/OSGeo/gdal/pull/9188 and I get compiler warnings, related to doing things like ``*(int64_t*)(&float_number) = ``

https://github.com/OSGeo/gdal/actions/runs/7775340143/job/21201210205?pr=9188

In GDAL, we use things like:
```
/** Byte-swap a 64 bit pointer */
#define CPL_SWAP64PTR(x)                                                       \
    do                                                                         \
    {                                                                          \
        uint64_t _n64;                                                         \
        void *_lx = x;                                                         \
        memcpy(&_n64, _lx, 8);                                                 \
        CPL_STATIC_ASSERT_IF_AVAILABLE(sizeof(*(x)) == 1 ||                    \
                                       sizeof(*(x)) == 8);                     \
        _n64 = CPL_SWAP64(_n64);                                               \
        memcpy(_lx, &_n64, 8);                                                 \
    } while (0)

/** Byte-swap a 64 bit pointer */
#define CPL_SWAPDOUBLE(p) CPL_SWAP64PTR(p)
```

**Comment by rouault** [comment]
(the good new though is that this PR passes the GDAL non regression tests on big-endian s390x)

**Comment by thbeu** [comment]
> and I get compiler warnings

Yes, I am aware of this and relaxed the strict types check by https://github.com/OSGeo/shapelib/pull/94/commits/3b6a54838d29ca8a36b42e7d66a934546f9ec0dc for that reason.

Having memcpy twice per swap is the type-safe variant but can lead to some performance degrading. I will make it an optional flag.

**Comment by rouault** [comment]
> Having memcpy twice per swap is the type-safe variant but can lead to some performance degrading. I will make it an optional flag.

A reasonable compiler in optimized mode will remove those memcpy(). Demo
```
$ cat test.c

#include <string.h>
#include <stdint.h>

double byte_swap_compliant(double d)
{
    uint64_t n;
    memcpy(&n, &d, sizeof(d));
    n = __builtin_bswap64(n);
    memcpy(&d, &n, sizeof(d));
    return d;
}

double byte_swap(double d)
{
    uint64_t n = __builtin_bswap64(*(uint64_t*)&d);
    return *(double*) &n;
}

$ gcc test.c -O2 -c && objdump -d test.o
[...]
0000000000000000 <byte_swap_compliant>:
   0:	f3 0f 1e fa          	endbr64 
   4:	66 48 0f 7e c0       	movq   %xmm0,%rax
   9:	48 0f c8             	bswap  %rax
   c:	66 48 0f 6e c0       	movq   %rax,%xmm0
  11:	c3                   	retq   
  12:	66 66 2e 0f 1f 84 00 	data16 nopw %cs:0x0(%rax,%rax,1)
  19:	00 00 00 00 
  1d:	0f 1f 00             	nopl   (%rax)

0000000000000020 <byte_swap>:
  20:	f3 0f 1e fa          	endbr64 
  24:	66 48 0f 7e c0       	movq   %xmm0,%rax
  29:	48 0f c8             	bswap  %rax
  2c:	66 48 0f 6e c0       	movq   %rax,%xmm0
  31:	c3                   	retq   
```

Same with clang.

Actually MSVC is less clever: https://godbolt.org/z/WjWM9Gbv7

The most sensitive performance critical parts are for byte swapping on X,Y,Z,M components of geometries, and that's only done on big-endian hosts (and for those perhaps also gcc/clang do a good job at optimizing the copies), for which I don't think we should care much about. I don't think the couple extra register moves in byte swapping done on little-endian on MSVC is going to be noticed at all, hence my preference for aliasing compliant code.

**Comment by thbeu** [comment]
There are now locations where memcpy is called three times just to obtain a value. This should be optimized.

**Comment by rouault** [comment]
hum, I've regiven a try at that PR updated with your last commit in GDAL, and this now crashes at runtime on s390x big-endian in a test involving reading a polygon shapefile: https://app.travis-ci.com/github/OSGeo/gdal/jobs/617327526 . I don't see anything suspicious in your last commit though... I was wondering about the potential lack of parenthesis in the macro definition around the ``x`` parameter (generally a good idea to add parenthesis), but should be safe here.

**Comment by rouault** [comment]
> Not sure what to do next actually. 😕

There's no error message. It just crashes somewhere in the middle of a test

**Comment by thbeu** [comment]
Did it also chrash without the last commit https://github.com/OSGeo/shapelib/pull/94/commits/befceca3c1578de2677299897d2a12d7f51c4ccd applied?

**Comment by thbeu** [comment]
OK, there was one more finding.

**Comment by rouault** [comment]
OK, GDAL Continous Integration is happy with the latest commit

**Comment by thbeu** [comment]
Shall I rename SHP_SWAPDOUBLE_2 to SHP_SWAPDOUBLE_CPY?

**Comment by rouault** [comment]
> Shall I rename SHP_SWAPDOUBLE_2 to SHP_SWAPDOUBLE_CPY?

yes, sounds good

**Comment by rouault** [comment]
ok, there was a typo breaking big endian compilation for which I've pushed a fix in your branch. All looks good to me and CI

**Comment by thbeu** [comment]
> ok, there was a typo breaking big endian compilation for which I've pushed a fix in your branch. All looks good to me and CI

Thanks for fixing. It's squashed.

**Comment by rouault** [comment]
> Is there any chance for a new release soon?

Let's things settle down a bit. Not before a couple months at least for a new release. Sorry shapelib is rather a slow moving project. It has never had such a burst of activity in the last 10+ years.

**Pull request merged**